### DISCUSSION TITLE
Libre 1 - Sensor age has not advanced 5427

### DISCUSSION THREAD

**Discussion by JohhnyP** [NightscoutFoundation/xDrip#2760] [Q&A]
Greetings.

I'm facing occasionally this error and cannot get readings from Bubble to xDrip.

I couldn't find any workaround it either fix itself after a while or I'm forced  to change sensor.

Manual readings through reader or phone scans work fine so I'm pretty sure this is an xDrip issue and not a sensor issue.

Anyone has any idea?

I can provide whatever kind of info you need.

I'm on latest xDrip and the master phone is Samsung Galaxy A40 with Android 11.

![20230401_163723](https://user-images.githubusercontent.com/11662759/229292638-8344539d-8977-4a0c-9123-7a6ff58ea38d.jpg)

Codeblock where the error is raised

```
else if (thisSensorAge == sensorAge) {
                // This is only a problem if we don't have a recent reading. It could happen that we have a recent
                // reading, and then BT disconnects and connects after a few seconds, and we have a new reading (where 
                // sensor did not advance). This does not mean that a sensor is not advancing. Only if this is happening
                // for a few minutes, this is a problem.
                if (BgReading.getTimeSinceLastReading() > 11 * 60 * 1000) {
                    Log.wtf(TAG, "Sensor age has not advanced: " + sensorAge);
                    JoH.static_toast_long(gs(R.string.sensor_clock_has_not_advanced));
                    Pref.setBoolean("nfc_age_problem", true);
                }
                return; // do not try to insert again
```

**Comment by nigelkeirross** [comment]
I am having similar  problems with my current Libre2 sensor, reading direct (OOP2) with no bridge. Previous experience has been good, but now I am only getting about 60% of the readings, (sometimes will go over an hour with no readings) often messages "Sensor age has not advanced" or "sensor age has gone backwards". In system status the BT to the sensor is always connected. This also happens when I try NFC read from xDrip+ - only about 50% "work" even tho always "Scanned OK!" message. Are there any settings to override, or is it really a poorly sensor?

**Comment by Navid200** [comment]
@psonnera

**Comment by tzachi-dar** [comment]
you seem to have errors on your readings. See the message Librepoint.iserror returning true. This is why xDrip is ignoring this readings.
You might see some more values under "show libre trend" but the values can not be trusted.

**Reply by nigelkeirross** [reply]
I also have similar errors along with my sensor age messages. Does this mean that the Linbre2 sensor is intermittently faulty?
I have tried to stich some log images together..
![logs](https://user-images.githubusercontent.com/21008320/232105451-545f558d-b69d-403d-8849-e0abfce3fdb9.jpg)

**Reply by psonnera** [reply]
Look at the rawSensorValue. So many zeroes... is that sensor really working?

**Reply by nigelkeirross** [reply]
@psonnera The sensor seems to be intermittent. Over the last few hours it has been almost normal - see attached xDrip+ graph. But I am still trying to understand are the gaps due to a bad sensor, or connectivity issues - as i still am relatively new to troubleshooting the direct bridgeless Libre2 operation. 
Also, it does seem that the historical gaps get filled in over time - I'm not sure if that is from the occasional NFC scans I do or just the BT readings. 
![graph](https://user-images.githubusercontent.com/21008320/232124180-e2353879-8ade-4237-8c49-3711f28150b5.png)

**Reply by psonnera** [reply]
@nigelkeirross if you scan manually and there's no backfilling I'd suspect the sensor. Your log shows 1 minute reading so I'm not convinced it's connectivity issues. You might have more insights on what's going on enabling raw data visualization https://xdrip.readthedocs.io/en/latest/use/display/#raw-data

**Reply by nigelkeirross** [reply]
Thanks, I already have that set, ut as you can see no such display. Could there be another setting that is conflicting to prevent the raw data display?

**Reply by psonnera** [reply]
Ooops https://xdrip.readthedocs.io/en/latest/use/lesscommon/#advanced-settings-for-libre-2

**Reply by nigelkeirross** [reply]
Was that Ooops or Ooops2? I already had both of those Libre2 settings turned on, yet still no Raw data. My xDrip is from 2022-08-03.

**Reply by psonnera** [reply]
Ooops was: sorry forgot to tell you to enable raw data from Libre 2 in less common settings

**Reply by psonnera** [reply]
Strange...

**Reply by nigelkeirross** [reply]
It has been mostly good for the last 12 hours until 57 minutes ago ~ 1800 UK time, and no readings since then. I've tried NFC about 12 times and still nothing. I only has 5 "Sensor age has gone backwards at 1800, 1802, 1804, 1808 & 1813 - none since then (about 3 mins ago). So mostly raw=0 values I assume. But I have seen recently 2 non-zero LibreOOPAlgorithm message:
1854 Creating BG time=7631 raw=2085
1852 Creating BG time = 7629 raw = 2131
1848 Creating BG time = 7625 raw = 2090
but almost all the other such messages are raw = 0.
So why is it not using one of the non-zero values?

**Reply by nigelkeirross** [reply]
@psonnera The NFC readings, all of which are failing, seem to have a pattern of -16 values for the last 3...
![NFCbehaviour](https://user-images.githubusercontent.com/21008320/232245944-be39c205-1229-4e7b-af78-f078ecca626e.png)

**Comment by PhilTU** [comment]
"Was that Ooops or Ooops2? I already had both of those Libre2 settings turned on" - can you check this, I don't think settings related to Out of Process should be enabled for this configuration.

**Reply by psonnera** [reply]
Agree. If OOP1 was enabled by error we'd see the message: ignoring measurement due to external algorithm enabled. So, no, not the issue. The issue is sensor reading an error code (-16) instead of BG.

**Comment by filipedanunes** [comment]
I'm having the same exact problems for the last months using libre 2 direct. Do you have a y clue how to solve this problem?

**Reply by Navid200** [reply]
@PhilTU

**Reply by PhilTU** [reply]
I think this arises from scanning close to a BLE transmission, leading to odd values from the sensor.

The problem is compounded by the rogue values being processed (rather than rejected) which leads to odd values in the smoothed valur which are then rejected, losng otherwise good data.

But we should try to document exactly if it's repeatable. Does it happen without NFC scanning?

**Reply by nigelkeirross** [reply]
@PhilTU @Navid200 @psonnera I have also been seeing many "sensor age has gone backwards" messages, which correspond to the raw=0 every-minute readings. I have modified xDrip to send the 1-minute readings to my looper app, as well as each time the libre2 connection changes state - mostly on/off (0/19,8). These are then stored to a mysql database in the cloud. 

I have many concerns about the BLE connection dropping between sensor & phone with Libre2 direct. So for the last 4 sensors I have been running in tandem - my main looping phone connected by the more reliable Bubble (mini) bridge, and a backup phone connected direct to the same sensor. In general there seem to be no problems running the 2 paths at the same time. However, I have been seeing a high level of raw=0 values on the direct phone. 9.7% (7865/81083) of the 1-minute readings are 0. Following @PhilTu's comment about this being caused by an NFC scan coinciding with the BLE reading (in this case the NFC would be from the Bubble device & since I have started this logging exercise I no longer NFC scan the direct connection - apart from at startup), as I want to be able to measure how quickly connection is re-established by itself. As I am storing all the CGM 5-minute readings from the Bridge phone, and all the 1-minute readings on the direct on the other phone, I should be able to see if indeed the raw=0 readings happen if they happen around the time of the Bubble (NFC) readings. 

The other observation I have relates to the frequent disconnections of the direct link. 42% (4344/10234) of the re-connections immediately disconnect again (after 2-3 seconds) with no reading being transmitted. I would like to know if anyone else has seen this & knows if anything can be done about this.

Finally I have observed that many of the successful re-connections happen immediately before a 1-minute reading is transmitted. So it looks like when the sensor has a reading to send, it makes an attempt to connect. I am yet to do a detailed analysis of this & see how this might help ensure that connections are more rapidly re-established. I have lots of data to analyse. If anyone is interested I'd be very happy to send the datasets.

**Reply by PhilTU** [reply]
I did notice that it stays connected direct to L2 in favourable conditions, with over 2 hours lapsed since "last connected" in one instance. I assume this is expected behaviour with the Libre streaming values every minute and xDrip only reconnecting in the absence of received data.

I haven't looked closely at how it works though. 
![Screenshot_20240830-192530](https://github.com/user-attachments/assets/7ca3fe6e-3849-470d-99b0-e66c30052795)

**Reply by nigelkeirross** [reply]
@PhilTU The attempt to re-connect is triggered in the BT callback - when the connection changes state to disconnected.

**Reply by SimonPhilpott** [reply]
I get this a lot, sensor age going backwards rejects the value. I find that this can be fixed by pressing stop sensor, then resetting caliberations. 

But it's not always reliable. Any caliberations are generally held back until the next reading, if you don't get a reading because it's rejected the caliberations table is blank and this stops further readings from being used. 

OOP2 show values being collected. Just my opinion but rejecting the values based on sensor time is overzelous when using L2 direct.

**Reply by SimonPhilpott** [reply]
Here's an example. I lost connection for over an hour. I rescanned the libre2 via NFC to re-establish connection, it backfilled and then started to throw sensor age errors. OOP2 shows data being received. 
![Screenshot_20250508_190256_xDrip+](https://github.com/user-attachments/assets/5fe3fd68-421b-45d0-856b-5cb37dd7c0d6)
![Screenshot_20250508_190309_xDrip+](https://github.com/user-attachments/assets/7b4b7990-e054-46ac-b292-b5c8b8f45eec)
![Screenshot_20250508_190548_OOPalgorithm2](https://github.com/user-attachments/assets/e5e1b291-018c-45e4-a246-dd82d488f632)
